### PR TITLE
Fix timestamp and read receipt group placement on ThreadView

### PR DESCRIPTION
**Pull request by luixxiul**
| |Before|After|
|-|---------|------|
|Thread panel|Screenshot from 2023-03-09 04-59-00|Screenshot from 2023-03-09 04-59-05|
|Timestamp|Screenshot from 2023-03-09 04-59-32|Screenshot from 2023-03-09 04-59-54|
|Read receipts|Screenshot from 2023-03-09 05-02-45|Screenshot from 2023-03-09 05-02-55
|Read receipts|Screenshot from 2023-03-09 05-06-40|Screenshot from 2023-03-09 05-06-48|
|MessageActionBar|Screenshot from 2023-03-09 05-35-57|Screenshot from 2023-03-09 05-43-39|

Fixes https://github.com/vector-im/element-web/issues/24773
Fixes https://github.com/vector-im/element-web/issues/23867
Fixes https://github.com/vector-im/element-web/issues/24744
Follow-up to https://github.com/matrix-org/matrix-react-sdk/pull/10284
For https://github.com/vector-im/element-web/issues/23569
For https://github.com/vector-im/element-web/issues/23326

## Checklist

-   [ ] Tests written for new code (and old code if feasible)
-   [ ] Linter and other CI checks pass
-   [ ] Sign-off given on the changes (see [CONTRIBUTING.md](https://github.com/matrix-org/matrix-react-sdk/blob/develop/CONTRIBUTING.md))

### CODE REVIEW

File: src/components/views/rooms/EventTile.tsx [hunk 1]
```diff
@@ -1197,8 +1197,8 @@ export class UnwrappedEventTile extends React.Component<EventTileProps, IState>
                             <a href={permalink} onClick={this.onPermalinkClicked}>
                                 {timestamp}
                             </a>
-                            {msgOption}
                         </div>,
+                        msgOption,
```
Comment (luixxiul) [inline src/components/views/rooms/EventTile.tsx:1201]: See: https://github.com/matrix-org/matrix-react-sdk/pull/10335/files#diff-9ff5535291b7c2e655ccd3d02c4f632b6f2dc9c30072514641e31c2a8751c57eR1287 and https://github.com/matrix-org/matrix-react-sdk/pull/10335/files#diff-9ff5535291b7c2e655ccd3d02c4f632b6f2dc9c30072514641e31c2a8751c57eR1398. Both of them are placed outside of `mx_EventTile_line`.

File: cypress/e2e/threads/threads.spec.ts [hunk 2]
```diff
@@ -199,16 +193,12 @@ describe("Threads", () => {
...
-            cy.get(".mx_EventTile_line .mx_EventTile_content")
-                // 76px: ThreadViewGroupSpacingStart + 14px + 6px
-                // 14px: avatar width
-                // See: _EventTile.pcss
-                .should("have.css", "margin-inline-start", "76px");
-            cy.get(".mx_EventTile_line")
-                // Make sure the margin is NOT applied to mx_EventTile_line
+        cy.get(".mx_ThreadView .mx_GenericEventListSummary[data-layout=bubble]").within(() => {
+            cy.get(".mx_EventTile_info.mx_EventTile_last")
+                // 49px: --EventTile_bubble-margin-inline-start
```
Comment (luixxiul) [inline cypress/e2e/threads/threads.spec.ts:198]: https://github.com/matrix-org/matrix-react-sdk/pull/10335/files#diff-5d3e25eb1d22dd7a96e0e87ea9c1298dc7dbce8eb965d30b2d1ebc1436bfec3fR1108

File: res/css/views/rooms/_EventTile.pcss [hunk 3]
```diff
@@ -1032,20 +1032,14 @@ $left-gutter: 64px;
     --ThreadView_group_spacing-end: 8px; /* same as padding */
 
     .mx_EventTile {
-        display: flex;
```
Comment (luixxiul) [inline res/css/views/rooms/_EventTile.pcss:1035]: https://github.com/matrix-org/matrix-react-sdk/pull/10335/files#diff-5d3e25eb1d22dd7a96e0e87ea9c1298dc7dbce8eb965d30b2d1ebc1436bfec3fR1105

Applied to bubble layout only

File: res/css/views/rooms/_EventTile.pcss [hunk 4]
```diff
@@ -1032,20 +1032,14 @@ $left-gutter: 64px;
...
-        flex-direction: column;
```
Comment (luixxiul) [inline res/css/views/rooms/_EventTile.pcss:1036]: On bubble layout flex-direction: row, the default declaration, is applied.

File: res/css/views/rooms/_EventTile.pcss [hunk 5]
```diff
@@ -1082,28 +1077,58 @@ $left-gutter: 64px;
...
                 }
+
+                .mx_ReadReceiptGroup {
+                    /* Move read receipt group inside ThreadView */
+                    inset-inline-end: 0;
+                }
+
+                &[data-self="true"] {
+                    /* Align the avatar and mx_EventTile_line at the center on hidden event tile line */
+                    align-items: center;
```
Comment (luixxiul) [inline res/css/views/rooms/_EventTile.pcss:1130]: https://github.com/matrix-org/matrix-react-sdk/pull/10335/files#diff-5d3e25eb1d22dd7a96e0e87ea9c1298dc7dbce8eb965d30b2d1ebc1436bfec3fR1235

File: res/css/views/rooms/_EventTile.pcss [hunk 6]
```diff
@@ -1082,28 +1077,58 @@ $left-gutter: 64px;
...
+                    /* Apply the same value as the main timeline */
+                    inset-inline: calc(-1 * var(--EventTile_bubble-margin-inline-start)) 0;
                 }
 
-                &.mx_EventTile_selected .mx_EventTile_line,
                 .mx_EventTile_line {
-                    .mx_EventTile_content,
-                    .mx_RedactedBody {
-                        margin-inline-start: calc(var(--ThreadView_group_spacing-start) + 14px + 6px);
+                    /* Not available on the actual UI anyway as the hidden event expand button is hidden above */
```
Comment (luixxiul) [inline res/css/views/rooms/_EventTile.pcss:1117]: https://github.com/matrix-org/matrix-react-sdk/pull/10335/files#diff-5d3e25eb1d22dd7a96e0e87ea9c1298dc7dbce8eb965d30b2d1ebc1436bfec3fR1048

**Comment by luixxiul** [comment]
I've found the style rules for read receipt group can be merged with that on `_TimelineCard.pcss`. Working on it.

File: res/css/views/right_panel/_TimelineCard.pcss [hunk 7]
```diff
@@ -170,16 +170,12 @@ limitations under the License.
...
     }
 
-    .mx_GenericEventListSummary_unstyledList, /* RR next to a message on the event list summary */
+    .mx_GenericEventListSummary_unstyledList,
     .mx_RoomView_MessageList {
-        /* RR next to a message on the messsge list */
         .mx_EventTile[data-layout="bubble"] {
+            /* RR next to a message on the event list summary and the messsge list */
             .mx_ReadReceiptGroup {
-                /* 6px: scroll bar width (magic number) */
```
Comment (luixxiul) [inline res/css/views/right_panel/_TimelineCard.pcss:178]: It looks that the magic number should in fact have been `var(--BaseCard_padding-inline)`.

File: res/css/views/rooms/_ReadReceiptGroup.pcss [hunk 8]
```diff
@@ -14,9 +14,12 @@ See the License for the specific language governing permissions and
 limitations under the License.
 */
 
-.mx_ReadReceiptGroup {
+/* Required against alphabetic order of CSS import */
+:root {
```
Comment (luixxiul) [inline res/css/views/rooms/_ReadReceiptGroup.pcss:18]: Otherwise this variable could not be applied to `res/css/views/right_panel/_BaseCard.pcss` (and any CSS files imported before this file).

File: res/css/views/rooms/_ReadReceiptGroup.pcss [hunk 9]
```diff
@@ -14,9 +14,12 @@ See the License for the specific language governing permissions and
 limitations under the License.
 */
 
-.mx_ReadReceiptGroup {
+/* Required due to alphabetic order of CSS import */
+:root {
```
Comment (luixxiul) [inline res/css/views/rooms/_ReadReceiptGroup.pcss:18]: Otherwise this variable could not be used on `res/css/views/right_panel/_BaseCard.pcss` (and any CSS files imported before this file).

**Comment by luixxiul** [comment]
Found some other things to be fixed... working on them.

File: res/css/views/right_panel/_TimelineCard.pcss [hunk 10]
```diff
@@ -170,16 +170,12 @@ limitations under the License.
...
-                /* stylelint-disable-next-line declaration-colon-space-after */
-                inset-inline-end: calc(
-                    -1 * var(--ReadReceiptGroup_EventBubbleTile-spacing-end) + $container-gap-width + 6px
-                );
+                inset-inline-end: var(--BaseCard_EventTile_bubble_RR-inset-inline-end);
```
Comment (luixxiul) [inline res/css/views/right_panel/_TimelineCard.pcss:178]: See: https://github.com/matrix-org/matrix-react-sdk/pull/10335/files#diff-5d3e25eb1d22dd7a96e0e87ea9c1298dc7dbce8eb965d30b2d1ebc1436bfec3fR1230

**Comment by andybalaam** [comment]
I'm looking, but it's going to take me a while to get my head around all this, next week.

**Comment by luixxiul** [comment]
No problem. It is a little bit large one and you seem to be busy for other tasks, so take your time :+1:

File: res/css/views/rooms/_EventTile.pcss [hunk 11]
```diff
@@ -1032,77 +1032,106 @@ $left-gutter: 64px;
     --ThreadView_group_spacing-end: 8px; /* same as padding */
 
     .mx_EventTile {
-        display: flex;
-        flex-direction: column;
```
Comment (richvdh) [inline res/css/views/rooms/_EventTile.pcss:1036]: why is it safe to remove these?

File: res/css/views/rooms/_EventTile.pcss [hunk 12]
```diff
@@ -1032,77 +1032,106 @@ $left-gutter: 64px;
...
         }
 
         /* handling for hidden events (e.g reactions) in the thread view */
         &.mx_EventTile_info {
             .mx_EventTile_avatar {
-                position: absolute;
-                top: 1.5px; /* Align with hidden event content */
-                margin-top: 0;
-                margin-bottom: 0;
+                margin-block: 0;
```
Comment (richvdh) [inline res/css/views/rooms/_EventTile.pcss:1042]: I have to say, I find the separate `margin-top` and `margin-bottom` properties easier to reason about. Is there a particular reason to prefer `margin-block`?

File: res/css/views/rooms/_EventTile.pcss [hunk 13]
```diff
@@ -1032,77 +1032,106 @@ $left-gutter: 64px;
...
                 width: 14px; /* avatar img size */
                 height: 14px; /* avatar img size */
             }
 
-            .mx_ViewSourceEvent_toggle {
-                display: none; /* hide the hidden event expand button, not enough
-                space, view source can still be used */
-            }
-
```
Comment (richvdh) [inline res/css/views/rooms/_EventTile.pcss:1057]: is moving this necessary?

File: src/components/views/rooms/EventTile.tsx (hunk 1 above)
Comment (richvdh) [reply src/components/views/rooms/EventTile.tsx:1201]: I didn't understand what these links were supposed to tell me. For the benefit of future reviewers:

This brings `TimelineRenderingType.Thread` in line with the other timeline rendering types by moving the `msgOption` (ie, the read receipt group) outside the `mx_EventTile_line`.

File: res/css/views/rooms/_EventTile.pcss [hunk 14]
```diff
@@ -1178,10 +1203,12 @@ $left-gutter: 64px;
...
 
         &[data-layout="bubble"] {
-            margin-inline: var(--BaseCard_EventTile-spacing-inline);
+            margin-inline-start: var(--BaseCard_EventTile-spacing-inline);
+            margin-inline-end: var(--EventTile_bubble-margin-inline-end); /* Reserve space for read receipt group */
 
             &::before {
-                inset-inline: calc(-1 * var(--BaseCard_EventTile-spacing-inline));
+                inset-inline-start: calc(-1 * var(--BaseCard_EventTile-spacing-inline));
+                inset-inline-end: calc(-1 * var(--EventTile_bubble-margin-inline-end));
```
[the remark this answers is not in the document]
Comment (richvdh) [reply res/css/views/rooms/_EventTile.pcss:1215]: please don't just link to another line in the diff. What am I supposed to learn by looking at that link?

File: res/css/views/rooms/_EventTile.pcss [hunk 15]
```diff
@@ -1234,8 +1234,6 @@ $left-gutter: 64px;
             }
 
             &[data-self="true"] {
-                align-items: flex-end;
```
[the remark this answers is not in the document]
Comment (richvdh) [reply res/css/views/rooms/_EventTile.pcss:1199]: again, what do I learn by looking at that link?

File: res/css/views/right_panel/_TimelineCard.pcss (hunk 10 above)
Comment (richvdh) [reply res/css/views/right_panel/_TimelineCard.pcss:178]: > See: https://github.com/matrix-org/matrix-react-sdk/pull/10335/files#diff-5d3e25eb1d22dd7a96e0e87ea9c1298dc7dbce8eb965d30b2d1ebc1436bfec3fR1230

Once more, just linking to a diff isn't very helpful.

In this case, this is a link to the other place that `--BaseCard_EventTile_bubble_RR-inset-inline-end` is now used.

**Review by richvdh — CHANGES_REQUESTED** [review]
I've started having a look at this, but there are quite a few things I don't understand. I have asked some questions below, though there are parts I haven't even looked at yet.

The changes have the air of being sensible, but they are hard to follow, and generally I wonder if this trying to fix too many things at once. Would it be possible to break this PR down into smaller increments to help us review it?

**Comment by luixxiul** [comment]
Thanks for the educational insights.

converting for now to make commits understandable for those who are not familiar with the style rules around EventTiles
